src/k8s/pkg/k8sd/api/certificates_refresh.go

```diff
@@ -165,56 +171,29 @@ func refreshCertsRunWorker(s state.State, r *http.Request, snap snap.Snap) respo
 			}
 
 			for {
+				err := client.WatchCertificateSigningRequest(
+					ctx,
+					csr.name,
+					func(request *certificatesv1.CertificateSigningRequest) (bool, error) {
+						return verifyCSRAndSetPKI(request, keyPEM, csr.certificate, csr.key)
+					})
+
+				if err == nil {
+					return nil
+				}
+
+				// Check if error is non-recoverable
+				if errors.Is(err, errInvalidCSR) {
+					return fmt.Errorf("certificate signing request failed: %w", err)
+				}
```

Comment:
better to use a `retry bool, err error` signature for the result, rather than communicate this with a "typed" error. there is currently nothing that prevents someone to change the signature of the error, and then this if-check breaks.